### DISCUSSION TITLE
Can't connect with remote ClickHouse database via self-hosted kubernetes

### DISCUSSION THREAD

**Discussion by richturnerdev** [plausible/analytics#1192] [Self-Hosted Support] [answered]
We wish to deploy a self-hosted Plausible service via kubernetes for a POC (we're happy with data not persisting - just looking to see if it works first). To do this, we've deployed a remote ClickHouse database also via kubernetes. As per the docs, we just need to plug in the clickhouse database url - however, I'm struggling to figure out the correct endpoint to use to connect successfully. We have a host that we're able to make http queries against and create databases with, but if I use `*host*/*database_name*` as the `CLICKHOUSE_DATABASE_URL` (similar to what is [in the docs](https://plausible.io/docs/self-hosting-configuration#database)) I am getting 404's in the logs. Any idea the correct url structure or if I'm missing something else?

`[error] Clickhousex.Protocol (#PID<0.178.0>) failed to connect: ** (ErlangError) Erlang error: "404 page not found"`

**Comment by ukutaht** [comment, marked as answer]
We use it just as you described - `CLICKHOUSE_DATABASE_URL=http://<username>:<password>@<host>:<port>/<database_name>`. That should work fine. If it doesn't, I would guess it's a connection issue between the two servers?

**Reply by richturnerdev** [reply]
Thanks @ukutaht. I have since recreated the error locally. I can start docker compose (using the supplied plausible template) and connect to the ClickHouse database by relying on the default env values (i.e. not specifying CLICKHOUSE_DATABASE_URL). However, as soon as I specify `CLICKHOUSE_DATABASE_URL=http://localhost:8123/plausible_events_db` I get the same connection errors as when I try remotely on Kubernetes (despite seeing the database when running this command: `curl 'http://localhost:8123/?query=show%20databases'` - see below)

<img width="395" alt="Screenshot 2021-07-19 at 16 09 14" src="https://user-images.githubusercontent.com/44242926/126188583-4cc98b0e-c257-4982-848c-27244534900b.png">

What CLICKHOUSE_DATABASE_URL would I use to get it working locally? I have a feeling this will help me construct the remote url as well. I have also tried adding a username and password to the ClickHouse database and doing `http://foo:bar@localhost:8123/plausible_events_db` to no avail.

**Reply by ukutaht** [reply]
⚠️ Warning: I haven't tried these things out and I'm not sure at all whether all of this is correct

I think the issue with `http://localhost:8123/plausible_events_db` is that the Clickhouse database does not have any ports exposed so it's only accessible through the Docker Compose network. If you want to access the database from your host machine, you need to add a port mapping to the [docker-compose file](https://github.com/plausible/hosting/blob/master/docker-compose.yml#L14) as follows:

```yml
 plausible_events_db:
    image: yandex/clickhouse-server:21.3.2.5
    ports:
      - 8123:8123
    ...
```

**Reply by richturnerdev** [reply]
Yes I already did that in order to be able to use the curl command to interact with it (in the same manner as the remote database), so the port is exposed.

**Reply by ukutaht** [reply]
Oh of course.

I'm not sure what's going on here. The error message is extra confusing `[error] Clickhousex.Protocol (#PID<0.178.0>) failed to connect: ** (ErlangError) Erlang error: "404 page not found"`. A 404 would mean that a the http connection was made succesfully and a 404 response was returned. That leads me to believe that the connection is working but there could be some sort of credentials issue or a version mismatch. Really not sure, just throwing stuff against the wall at this point.

What version of Clickhouse are you running? Plausible has only been tested up to `21.3.2.5`

**Reply by richturnerdev** [reply]
Yes using Clickhouse `21.3.2.5` both locally and remotely. Just strange that it works locally without the url env variable. Does it just connect via the network when you're not specifying the url? Are you able to replicate the issue? As I say, I'm just running it locally with the official docker compose template provided.

**Reply by richturnerdev** [reply]
If I use the url (`http://localhost:8123/plausible_events_db`) in the browser I get the following message:

```
There is no handle /plausible_events_db

Use / or /ping for health checks.
Or /replicas_status for more sophisticated health checks.

Send queries from your program with POST method or GET /?query=...

Use clickhouse-client:

For interactive data analysis:
    clickhouse-client

For batch query processing:
    clickhouse-client --query='SELECT 1' > result
    clickhouse-client < query > result
```

Which is why it makes me think it's a syntax issue with the url.
